src/modules/musig/main_impl.h

```diff
@@ -449,196 +475,195 @@ int secp256k1_musig_partial_signature_parse(const secp256k1_context* ctx, secp25
     secp256k1_scalar sk;
-    secp256k1_scalar e, k;
+    secp256k1_ge pk;
+    secp256k1_scalar e, b, k[2];
+    secp256k1_scalar mu;
+    int combined_nonce_parity;
 
     VERIFY_CHECK(ctx != NULL);
     ARG_CHECK(partial_sig != NULL);
-    ARG_CHECK(session != NULL);
-    ARG_CHECK(session->magic == session_magic);
-    ARG_CHECK(session->round == 2);
-    ARG_CHECK(session->has_secret_data);
+    ARG_CHECK(secnonce != NULL);
+    ARG_CHECK(keypair != NULL);
+    ARG_CHECK(pre_session != NULL);
+    ARG_CHECK(pre_session->magic == pre_session_magic);
+    ARG_CHECK(session_cache != NULL);
+    {
+        /* Check in constant time if secnonce has been zeroed. */
+        size_t i;
+        unsigned char secnonce_acc = 0;
+        for (i = 0; i < sizeof(*secnonce); i++) {
+            secnonce_acc |= secnonce->data[i];
+        }
+        secp256k1_declassify(ctx, &secnonce_acc, sizeof(secnonce_acc));
+        ARG_CHECK(secnonce_acc != 0);
+    }
 
-    /* build message hash */
-    secp256k1_musig_compute_messagehash(ctx, msghash, session);
-    secp256k1_scalar_set_b32(&e, msghash, NULL);
+    secp256k1_musig_secnonce_load(k, secnonce);
+    memset(secnonce, 0, sizeof(*secnonce));
 
-    secp256k1_scalar_set_b32(&sk, session->seckey, &overflow);
-    if (overflow) {
-        secp256k1_scalar_clear(&sk);
+    /* Obtain the signer's public key point and determine if the sk is
+     * negated before signing. That happens if if the signer's pubkey has an odd
+     * Y coordinate XOR the MuSig-combined pubkey has an odd Y coordinate XOR
+     * (if tweaked) the internal key has an odd Y coordinate.
+     *
+     * This can be seen by looking at the sk key belonging to `combined_pk`.
+     * Let's define
+     * P' := mu_0*|P_0| + ... + mu_n*|P_n| where P_i is the i-th public key
+     * point x_i*G, mu_i is the i-th musig coefficient and |.| is a function
+     * that normalizes a point to an even Y by negating if necessary similar to
+     * secp256k1_extrakeys_ge_even_y. Then we have
+     * P := |P'| + t*G where t is the tweak.
+     * And the combined xonly public key is
+     * |P| = x*G
+     *      where x = sum_i(b_i*mu_i*x_i) + b'*t
+     *            b' = -1 if P != |P|, 1 otherwise
+     *            b_i = -1 if (P_i != |P_i| XOR P' != |P'| XOR P != |P|) and 1
+     *                otherwise.
+     */
+    if (!secp256k1_keypair_load(ctx, &sk, &pk, keypair)) {
         return 0;
     }
-
-    secp256k1_scalar_set_b32(&k, session->secnonce, &overflow);
-    if (overflow || secp256k1_scalar_is_zero(&k)) {
-        secp256k1_scalar_clear(&sk);
-        secp256k1_scalar_clear(&k);
-        return 0;
+    secp256k1_fe_normalize_var(&pk.y);
+    if((secp256k1_fe_is_odd(&pk.y)
+            + pre_session->pk_parity
+            + (pre_session->is_tweaked
+                && pre_session->internal_key_parity))
+            % 2 == 1) {
```

Comment:
I thought the intent of this was to be a glorious `a || b || (c && d)`, but that's not the case.
I never saw such kind of condition with boolean as integers.

Reply:
even with the comment quite confusing :)